File: apps/android/app/src/main/java/com/exponential/app/ui/issue/AgentPlanPanel.kt
package com.exponential.app.ui.issue

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowCircleUp
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.QuestionMark
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.SmartToy
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.exponential.app.data.db.IssueEventEntity
import com.exponential.app.data.db.UserEntity
import com.exponential.app.ui.markdown.MarkdownView
import com.exponential.app.ui.theme.PlanColors
import com.exponential.app.ui.theme.glassSection
import kotlinx.coroutines.launch

// Agent lifecycle events shown in the quiet activity feed (and used to detect a
// terminal error for the Retry affordance). Mirrors AGENT_EVENT_TYPES in
// apps/web/src/components/agent-plan-panel.tsx.
internal val agentEventTypes = setOf(
    "agent_started", "plan_ready", "agent_question",
    "agent_answer", "pr_opened", "pr_merged", "agent_error",
)

private val PanelText = Color.White.copy(alpha = 0.9f)
private val PanelMeta = Color.White.copy(alpha = 0.5f)
private val PanelFieldBg = Color.White.copy(alpha = 0.06f)
private val PanelAccent = Color(red = 0.42f, green = 0.64f, blue = 1.0f)
private val ErrorRed = Color(0xFFF87171)

// First-class panel for the agent plan/question lifecycle, replacing the
// plan/question comment rows (mirror of apps/web/src/components/agent-plan-panel.tsx).
// State is driven by the synced `issue` columns; the plan/question TEXT is
// fetched via agentPlan.getState (server-only, not in Electric).
@Composable
fun AgentPlanPanel(
    issueId: String,
    canApprovePlan: Boolean,
    viewModel: AgentPlanPanelViewModel = hiltViewModel(),
) {
    LaunchedEffect(issueId) { viewModel.bind(issueId) }
    val state by viewModel.state.collectAsStateWithLifecycle()
    val issue = state.issue
    val scope = rememberCoroutineScope()

    val agentEvents = remember(state.events) { state.events.filter { it.type in agentEventTypes } }
    val latestIsError = agentEvents.lastOrNull()?.type == "agent_error"
    val planState = issue?.agentPlanState

    // Render nothing when there is no agent involvement.
    if (issue == null || (planState == null && !latestIsError)) return

    var busy by remember { mutableStateOf<String?>(null) }
    var answer by remember { mutableStateOf("") }

    val finished = issue.status == "done" || issue.status == "cancelled"
    val implementing = !finished && planState == "approved" && issue.prState == null && !latestIsError

    fun act(label: String, block: suspend () -> Unit) {
        scope.launch {
            busy = label
            try {
                block()
            } finally {
                // Always clear, even if the action throws or the coroutine is
                // cancelled — otherwise the buttons stay disabled forever.
                busy = null
            }
        }
    }

    Column(modifier = Modifier.fillMaxWidth().glassSection().padding(12.dp)) {
        // Header
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                Icons.Filled.SmartToy,
                contentDescription = null,
                modifier = Modifier.size(16.dp),
                tint = PlanColors.AwaitingApproval,
            )
            Spacer(Modifier.width(6.dp))
            Text("Agent plan", style = MaterialTheme.typography.labelLarge, color = PanelText)
            if (issue.agentPlanRevision > 0) {
                Spacer(Modifier.width(6.dp))
                Text("rev ${issue.agentPlanRevision}", style = MaterialTheme.typography.labelSmall, color = PanelMeta)
            }
            Spacer(Modifier.weight(1f))
            if (planState == "approved" && issue.agentPlanApprovedAt != null) {
                Icon(Icons.Filled.Check, contentDescription = null, modifier = Modifier.size(12.dp), tint = PlanColors.Approved)
                Spacer(Modifier.width(4.dp))
                Text("Approved", style = MaterialTheme.typography.labelSmall, color = PlanColors.Approved)
            }
        }

        Spacer(Modifier.height(8.dp))

        when (planState) {
            "drafting", "planning" -> LoadingRow("Agent is working on a plan…")
            "awaiting_answer" -> QuestionContent(
                question = state.questionText,
                canApprovePlan = canApprovePlan,
                answer = answer,
                onAnswerChange = { answer = it },
                sending = busy == "answer",
                enabled = busy == null,
                onSend = {
                    val text = answer.trim()
                    if (text.isNotEmpty()) act("answer") { viewModel.answerQuestion(text); answer = "" }
                },
            )
            "awaiting_approval", "approved" -> PlanContent(
                planText = state.planText,
                showApproval = planState == "awaiting_approval" && canApprovePlan,
                busy = busy,
                onApprove = { act("approve") { viewModel.approvePlan() } },
                onRequestChanges = { act("request") { viewModel.requestChanges() } },
                implementing = implementing,
            )
            else -> {}
        }

        if (latestIsError) {
            Spacer(Modifier.height(8.dp))
            ErrorBanner(
                canRetry = canApprovePlan,
                retrying = busy == "retry",
                enabled = busy == null,
                onRetry = { act("retry") { viewModel.retry() } },
            )
        }
    }
}

@Composable
private fun LoadingRow(text: String) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        CircularProgressIndicator(modifier = Modifier.size(14.dp), strokeWidth = 2.dp, color = PlanColors.AwaitingApproval)
        Text(text, style = MaterialTheme.typography.bodySmall, color = PanelMeta)
    }
}

@Composable
private fun QuestionContent(
    question: String?,
    canApprovePlan: Boolean,
    answer: String,
    onAnswerChange: (String) -> Unit,
    sending: Boolean,
    enabled: Boolean,
    onSend: () -> Unit,
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Icon(Icons.Filled.QuestionMark, contentDescription = null, modifier = Modifier.size(14.dp), tint = PlanColors.AwaitingAnswer)
        Spacer(Modifier.width(6.dp))
        Text("The agent has a question", style = MaterialTheme.typography.labelMedium, color = PlanColors.AwaitingAnswer)
    }
    Spacer(Modifier.height(6.dp))
    if (!question.isNullOrEmpty()) {
        MarkdownView(question)
    } else {
        Text("Loading…", style = MaterialTheme.typography.bodySmall, color = PanelMeta)
    }
    if (canApprovePlan) {
        Spacer(Modifier.height(8.dp))
        Row(verticalAlignment = Alignment.Bottom, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            BasicTextField(
                value = answer,
                onValueChange = onAnswerChange,
                enabled = enabled,
                textStyle = MaterialTheme.typography.bodyMedium.copy(color = PanelText),
                cursorBrush = SolidColor(PanelAccent),
                maxLines = 5,
                modifier = Modifier.weight(1f),
                decorationBox = { inner ->
                    Box(
                        modifier = Modifier
                            .clip(RoundedCornerShape(18.dp))
                            .background(PanelFieldBg)
                            .padding(horizontal = 12.dp, vertical = 10.dp),
                    ) {
                        if (answer.isEmpty()) {
                            Text("Answer the agent…", style = MaterialTheme.typography.bodyMedium, color = PanelMeta)
                        }
                        inner()
                    }
                },
            )
            IconButton(onClick = onSend, enabled = enabled && answer.isNotBlank()) {
                if (sending) {
                    CircularProgressIndicator(modifier = Modifier.size(20.dp), strokeWidth = 2.dp, color = PanelAccent)
                } else {
                    Icon(
                        Icons.Filled.ArrowCircleUp,
                        contentDescription = "Send answer",
                        modifier = Modifier.size(30.dp),
                        tint = if (answer.isBlank()) Color.White.copy(alpha = 0.3f) else PanelAccent,
                    )
                }
            }
        }
    }
}

@Composable
private fun PlanContent(
    planText: String?,
    showApproval: Boolean,
    busy: String?,
    onApprove: () -> Unit,
    onRequestChanges: () -> Unit,
    implementing: Boolean,
) {
    if (!planText.isNullOrEmpty()) {
        MarkdownView(planText)
    } else {
        Text("Loading plan…", style = MaterialTheme.typography.bodySmall, color = PanelMeta)
    }
    if (showApproval) {
        Spacer(Modifier.height(10.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(
                onClick = onApprove,
                enabled = busy == null,
                colors = ButtonDefaults.buttonColors(
                    containerColor = Color(0xFF22C55E).copy(alpha = 0.22f),
                    contentColor = Color(0xFF22C55E),
                ),
            ) {
                Icon(Icons.Filled.Check, contentDescription = null, modifier = Modifier.size(14.dp))
                Spacer(Modifier.width(6.dp))
                Text(if (busy == "approve") "Approving…" else "Approve")
            }
            OutlinedButton(onClick = onRequestChanges, enabled = busy == null) {
                Text(if (busy == "request") "Requesting…" else "Request changes")
            }
        }
    }
    if (implementing) {
        Spacer(Modifier.height(10.dp))
        LoadingRow("Agent is implementing the approved plan…")
    }
}

@Composable
private fun ErrorBanner(canRetry: Boolean, retrying: Boolean, enabled: Boolean, onRetry: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(ErrorRed.copy(alpha = 0.08f), RoundedCornerShape(8.dp))
            .border(0.5.dp, ErrorRed.copy(alpha = 0.25f), RoundedCornerShape(8.dp))
            .padding(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.size(6.dp).clip(CircleShape).background(ErrorRed))
        Spacer(Modifier.width(8.dp))
        Text("The agent hit an error.", style = MaterialTheme.typography.bodySmall, color = PanelMeta)
        if (canRetry) {
            Spacer(Modifier.weight(1f))
            OutlinedButton(onClick = onRetry, enabled = enabled) {
                Icon(Icons.Filled.Refresh, contentDescription = null, modifier = Modifier.size(14.dp))
                Spacer(Modifier.width(6.dp))
                Text(if (retrying) "Retrying…" else "Retry")
            }
        }
    }
}

// A quiet, collapsible feed of agent lifecycle events. Separate from the human
// comment thread so routine agent activity doesn't read as conversation.
// Mirror of apps/web/src/components/agent-activity-feed.tsx.
@Composable
fun AgentActivityFeed(events: List<IssueEventEntity>, usersById: Map<String, UserEntity>) {
    val agentEvents = remember(events) {
        events.filter { it.type in agentEventTypes }.sortedBy { it.createdAt }
    }
    if (agentEvents.isEmpty()) return
    var expanded by remember { mutableStateOf(false) }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(Color.White.copy(alpha = 0.03f), RoundedCornerShape(8.dp))
            .padding(10.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().clickable { expanded = !expanded },
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                if (expanded) Icons.Filled.KeyboardArrowDown else Icons.Filled.KeyboardArrowRight,
                contentDescription = null,
                modifier = Modifier.size(16.dp),
                tint = PanelMeta,
            )
            Spacer(Modifier.width(4.dp))
            Text("Agent activity (${agentEvents.size})", style = MaterialTheme.typography.labelMedium, color = PanelMeta)
        }
        if (expanded) {
            Spacer(Modifier.height(6.dp))
            agentEvents.forEach { event ->
                key(event.id) {
                    val who = usersById[event.actorUserId]?.let { it.name ?: it.email } ?: "Agent"
                    Row(
                        modifier = Modifier.fillMaxWidth().padding(vertical = 3.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        Box(Modifier.size(6.dp).clip(CircleShape).background(PanelMeta))
                        Text(
                            "$who ${agentEventVerb(event.type)} · ${relativeTime(event.createdAt)}",
                            style = MaterialTheme.typography.labelSmall,
                            color = PanelMeta,
                        )
                    }
                }
            }
        }
    }
}

// Human-readable verb for an issue/agent event type (shared with CommentThread's
// non-agent EventRow).
internal fun agentEventVerb(type: String): String = when (type) {
    "status_changed" -> "changed the status"
    "assignee_changed" -> "changed the assignee"
    "label_added" -> "added a label"
    "label_removed" -> "removed a label"
    "pr_opened" -> "opened a pull request"
    "pr_merged" -> "merged the pull request"
    "plan_ready" -> "posted a plan for review"
    "agent_error" -> "hit an error"
    "agent_started" -> "started working"
    "agent_question" -> "asked a question"
    "agent_answer" -> "answered the agent"
    else -> type.replace('_', ' ')
}

File: apps/ios/Exponential/UI/Issue/AgentPlanPanel.swift
import ExpCore
import ExpUI
import GRDB
import MarkdownUI
import SwiftUI

// Agent lifecycle events shown in the quiet activity feed (and used to detect a
// terminal error for the Retry affordance). Mirrors AGENT_EVENT_TYPES in
// apps/web/src/components/agent-plan-panel.tsx.
let agentEventTypes: Set<String> = [
    "agent_started", "plan_ready", "agent_question",
    "agent_answer", "pr_opened", "pr_merged", "agent_error",
]

/// First-class panel for the agent plan/question lifecycle, replacing the
/// plan/question comment rows (mirror of apps/web/src/components/agent-plan-panel.tsx).
/// State is driven by the synced `issue.agentPlanState`; the plan/question TEXT,
/// revision, and approval come from the synced `agent_runs` row.
struct AgentPlanPanel: View {
    let issue: IssueEntity
    let canApprovePlan: Bool

    @Environment(AppDependencies.self) private var deps
    @Environment(\.accountId) private var accountId

    @State private var agentRun: AgentRunEntity?
    @State private var answer = ""
    @State private var busy: PanelAction?
    @State private var events: [IssueEventEntity] = []
    @State private var observationTask: Task<Void, Never>?
    @State private var runObservationTask: Task<Void, Never>?
    @State private var showDiff = false

    private enum PanelAction: Equatable { case approve, requestChanges, answer, retry }

    private var agentEvents: [IssueEventEntity] {
        events.filter { agentEventTypes.contains($0.type) }
    }
    private var latestIsError: Bool { agentEvents.last?.type == "agent_error" }
    private var finished: Bool { issue.status == "done" || issue.status == "cancelled" }
    private var implementing: Bool {
        !finished && issue.agentPlanState == "approved" && issue.prState == nil && !latestIsError
    }
    // Plan/question text come from the synced `agent_runs` row (server-authored
    // jsonb {text}); unwrap them the same way as a comment body.
    private var planText: String? {
        let t = getCommentBodyText(agentRun?.planText)
        return t.isEmpty ? nil : t
    }
    private var questionText: String? {
        let t = getCommentBodyText(agentRun?.question)
        return t.isEmpty ? nil : t
    }

    var body: some View {
        Group {
            if issue.agentPlanState != nil || latestIsError || issue.prUrl != nil {
                VStack(alignment: .leading, spacing: 10) {
                    header
                    content
                    if latestIsError { errorBanner }
                    prSection
                }
                .padding(12)
                .glassSection()
            }
        }
        .onAppear { startObserving() }
        .onDisappear {
            observationTask?.cancel()
            runObservationTask?.cancel()
        }
    }

    // MARK: - Header

    private var header: some View {
        HStack(spacing: 6) {
            Image(systemName: "sparkles").font(.caption).foregroundStyle(Color.accentColor)
            Text("Agent plan").font(.subheadline.weight(.semibold)).foregroundStyle(.white)
            if let revision = agentRun?.planRevision, revision > 0 {
                Text("rev \(revision)")
                    .font(.caption2).foregroundStyle(.white.opacity(TextOpacity.tertiary))
            }
            Spacer()
            if issue.agentPlanState == "approved", agentRun?.approvedAt != nil {
                HStack(spacing: 4) {
                    Image(systemName: "checkmark.circle.fill").font(.caption2)
                    Text("Approved").font(.caption2.weight(.medium))
                }
                .foregroundStyle(.green)
            }
        }
    }

    // MARK: - State content

    @ViewBuilder
    private var content: some View {
        switch issue.agentPlanState {
        case "drafting", "planning":
            loadingRow("Agent is working on a plan…")
        case "awaiting_answer":
            questionContent
        case "awaiting_approval", "approved":
            planContent
        default:
            EmptyView()
        }
    }

    private func loadingRow(_ text: String) -> some View {
        HStack(spacing: 8) {
            ProgressView().controlSize(.small).tint(.white)
            Text(text).font(.caption).foregroundStyle(.white.opacity(TextOpacity.secondary))
        }
    }

    @ViewBuilder
    private var questionContent: some View {
        VStack(alignment: .leading, spacing: 8) {
            Label("The agent has a question", systemImage: "questionmark.circle")
                .font(.caption.weight(.medium)).foregroundStyle(.orange)
            if let q = questionText, !q.isEmpty {
                Markdown(q).markdownTheme(.gitHub).frame(maxWidth: .infinity, alignment: .leading)
            } else {
                Text("Loading…").font(.caption).foregroundStyle(.white.opacity(TextOpacity.tertiary))
            }
            if canApprovePlan {
                HStack(alignment: .bottom, spacing: 8) {
                    TextField("Answer the agent…", text: $answer, axis: .vertical)
                        .lineLimit(1...5)
                        .textFieldStyle(.plain)
                        .padding(.horizontal, 12).padding(.vertical, 8)
                        .background(Color.white.opacity(0.04))
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                        .foregroundStyle(.white)
                        .disabled(busy != nil)
                    Button {
                        let text = answer.trimmingCharacters(in: .whitespacesAndNewlines)
                        Task { await answerQuestion(text) }
                    } label: {
                        Group {
                            if busy == .answer {
                                ProgressView().controlSize(.mini).tint(.white)
                            } else {
                                Image(systemName: "paperplane.fill")
                            }
                        }
                        .padding(8)
                        .background(.blue, in: RoundedRectangle(cornerRadius: 8))
                        .foregroundStyle(.white)
                    }
                    .disabled(busy != nil || answer.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                }
            }
        }
    }

    @ViewBuilder
    private var planContent: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let p = planText, !p.isEmpty {
                Markdown(p).markdownTheme(.gitHub).frame(maxWidth: .infinity, alignment: .leading)
            } else {
                Text("Loading plan…").font(.caption).foregroundStyle(.white.opacity(TextOpacity.tertiary))
            }
            if issue.agentPlanState == "awaiting_approval", canApprovePlan {
                HStack(spacing: 8) {
                    Button { Task { await runAction(.approve) { try await deps.agentPlanApi.approvePlan(accountId: accountId, issueId: issue.id) } } } label: {
                        actionLabel("Approve", systemImage: "checkmark", loading: busy == .approve)
                            .foregroundStyle(.green)
                            .background(.green.opacity(0.18), in: RoundedRectangle(cornerRadius: 6))
                    }
                    .buttonStyle(.plain)
                    .disabled(busy != nil)
                    Button { Task { await runAction(.requestChanges) { try await deps.agentPlanApi.requestChanges(accountId: accountId, issueId: issue.id) } } } label: {
                        actionLabel("Request changes", systemImage: "pencil", loading: busy == .requestChanges)
                            .glassButton()
                    }
                    .buttonStyle(.plain)
                    .disabled(busy != nil)
                }
            }
            if implementing {
                loadingRow("Agent is implementing the approved plan…")
            }
        }
    }

    private func actionLabel(_ title: String, systemImage: String, loading: Bool) -> some View {
        HStack(spacing: 6) {
            if loading { ProgressView().controlSize(.mini).tint(.white) }
            else { Image(systemName: systemImage) }
            Text(title)
        }
        .font(.caption.weight(.semibold))
        .padding(.horizontal, 12).padding(.vertical, 6)
    }

    // The PR (branch + link) and an inline diff disclosure, shown once the agent
    // has opened a pull request. Mirrors the macOS panel's PR section.
    @ViewBuilder
    private var prSection: some View {
        if let prUrl = issue.prUrl, let url = URL(string: prUrl) {
            VStack(alignment: .leading, spacing: 6) {
                if let branch = issue.branch, !branch.isEmpty {
                    Text(branch)
                        .font(.caption2.monospaced())
                        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                }
                Link(destination: url) {
                    Label("View pull request", systemImage: "arrow.triangle.branch")
                        .font(.caption.weight(.medium))
                }
                .tint(Accent.indigo)
                DisclosureGroup("Changed files", isExpanded: $showDiff) {
                    DiffView(issueId: issue.id).padding(.top, 6)
                }
                .font(.caption)
                .tint(.white.opacity(TextOpacity.secondary))
            }
        }
    }

    @ViewBuilder
    private var errorBanner: some View {
        HStack(spacing: 8) {
            Image(systemName: "exclamationmark.triangle.fill").font(.caption).foregroundStyle(.red)
            Text("The agent hit an error.")
                .font(.caption).foregroundStyle(.white.opacity(TextOpacity.secondary))
            Spacer()
            if canApprovePlan {
                Button { Task { await runAction(.retry) { try await deps.agentPlanApi.retry(accountId: accountId, issueId: issue.id) } } } label: {
                    HStack(spacing: 6) {
                        if busy == .retry { ProgressView().controlSize(.mini).tint(.white) }
                        else { Image(systemName: "arrow.clockwise") }
                        Text(busy == .retry ? "Retrying…" : "Retry")
                    }
                    .font(.caption.weight(.medium))
                    .padding(.horizontal, 10).padding(.vertical, 6)
                    .glassButton()
                }
                .buttonStyle(.plain)
                .disabled(busy != nil)
            }
        }
        .padding(10)
        .background(Color.red.opacity(0.08))
        .overlay(RoundedRectangle(cornerRadius: 8).stroke(Color.red.opacity(0.25), lineWidth: 0.5))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }

    // MARK: - Actions

    private func runAction(_ action: PanelAction, _ fn: @escaping () async throws -> Void) async {
        busy = action
        defer { busy = nil }
        try? await fn()
    }

    private func answerQuestion(_ text: String) async {
        guard !text.isEmpty else { return }
        busy = .answer
        defer { busy = nil }
        do {
            try await deps.agentPlanApi.answerQuestion(accountId: accountId, issueId: issue.id, answer: text)
            answer = ""
        } catch {}
    }

    // MARK: - Event + agent-run observation

    // Both the activity feed and the plan/question text are driven entirely by
    // synced local rows now — no `agentPlan.getState` round-trip.
    private func startObserving() {
        observationTask?.cancel()
        runObservationTask?.cancel()
        guard let pool = try? deps.db.pool(forAccountId: accountId) else { return }
        observationTask = Task {
            let eventObs = ValueObservation.tracking { db in
                try IssueEventEntity
                    .filter(Column("issue_id") == issue.id)
                    .order(Column("created_at").asc)
                    .fetchAll(db)
            }
            do {
                for try await rows in eventObs.values(in: pool) { self.events = rows }
            } catch {}
        }
        runObservationTask = Task {
            let runObs = ValueObservation.tracking { db in
                try AgentRunEntity.fetchOne(db, key: issue.id)
            }
            do {
                for try await row in runObs.values(in: pool) { self.agentRun = row }
            } catch {}
        }
    }
}

/// A quiet, collapsible feed of agent lifecycle events. Separate from the human
/// comment thread so routine agent activity doesn't read as conversation.
/// Mirror of apps/web/src/components/agent-activity-feed.tsx.
struct AgentActivityFeed: View {
    let events: [IssueEventEntity]
    let users: [String: UserEntity]
    @State private var expanded = false

    private var agentEvents: [IssueEventEntity] {
        events.filter { agentEventTypes.contains($0.type) }
            .sorted { $0.createdAt < $1.createdAt }
    }

    var body: some View {
        if !agentEvents.isEmpty {
            VStack(alignment: .leading, spacing: 0) {
                Button { expanded.toggle() } label: {
                    HStack(spacing: 6) {
                        Image(systemName: expanded ? "chevron.down" : "chevron.right").font(.caption2)
                        Image(systemName: "bolt.horizontal").font(.caption2)
                        Text("Agent activity (\(agentEvents.count))").font(.caption)
                        Spacer()
                    }
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                if expanded {
                    VStack(alignment: .leading, spacing: 2) {
                        ForEach(agentEvents) { event in
                            HStack(spacing: 8) {
                                Circle().fill(.white.opacity(TextOpacity.tertiary)).frame(width: 6, height: 6)
                                Text("\(who(event)) \(eventPhrase(event, users: users, labels: nil))")
                                    .font(.caption).foregroundStyle(.white.opacity(TextOpacity.secondary))
                                Spacer()
                            }
                            .padding(.vertical, 3)
                        }
                    }
                    .padding(.top, 6)
                }
            }
            .padding(10)
            .background(Color.white.opacity(0.03))
            .clipShape(RoundedRectangle(cornerRadius: 8))
        }
    }

    private func who(_ event: IssueEventEntity) -> String {
        event.actorUserId.flatMap { users[$0] }.map { $0.name ?? $0.email } ?? "Agent"
    }
}

/// Human-readable verb for an issue/agent event type (the generic fallback used
/// when an event has no payload to render richly).
func agentEventVerb(_ type: String) -> String {
    switch type {
    case "status_changed": return "changed the status"
    case "assignee_changed": return "changed the assignee"
    case "label_added": return "added a label"
    case "label_removed": return "removed a label"
    case "pr_opened": return "opened a pull request"
    case "pr_merged": return "merged the pull request"
    case "plan_ready": return "posted a plan for review"
    case "agent_error": return "hit an error"
    case "agent_started": return "started working"
    case "agent_question": return "asked a question"
    case "agent_answer": return "answered the agent"
    default: return type.replacingOccurrences(of: "_", with: " ")
    }
}

/// Human label for an issue_status enum value.
func statusLabel(_ s: String) -> String {
    switch s {
    case "backlog": return "Backlog"
    case "todo": return "Todo"
    case "in_progress": return "In Progress"
    case "done": return "Done"
    case "cancelled": return "Cancelled"
    default: return s.replacingOccurrences(of: "_", with: " ").capitalized
    }
}

/// Pull a string or integer scalar out of an issue_event's JSON payload (stored
/// as stringified JSON). Returns nil for missing/null/empty values.
func eventField(_ payload: String?, _ key: String) -> String? {
    guard let payload, let data = payload.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let value = obj[key], !(value is NSNull) else { return nil }
    if let s = value as? String { return s.isEmpty ? nil : s }
    if let i = value as? Int { return String(i) }
    if let d = value as? Double { return String(Int(d)) }
    return nil
}

/// A rich activity phrase from the event type + payload (status from→to, PR #N,
/// assigned/unassigned, label name). Resolves user/label names when the maps are
/// supplied; falls back to the generic verb for events without a payload.
/// Mirrors the web activity timeline (and the Linux `eventPhrase`).
func eventPhrase(
    _ event: IssueEventEntity,
    users: [String: UserEntity],
    labels: [String: LabelEntity]?
) -> String {
    switch event.type {
    case "status_changed":
        guard let to = eventField(event.payload, "to") else { return "changed the status" }
        if let from = eventField(event.payload, "from") {
            return "changed status from \(statusLabel(from)) to \(statusLabel(to))"
        }
        return "changed status to \(statusLabel(to))"
    case "assignee_changed":
        guard let to = eventField(event.payload, "to") else { return "unassigned this issue" }
        if let name = users[to].map({ $0.name ?? $0.email }) {
            return "assigned \(name)"
        }
        return "assigned this issue"
    case "label_added":
        if let id = eventField(event.payload, "labelId"), let name = labels?[id]?.name {
            return "added label \(name)"
        }
        return "added a label"
    case "label_removed":
        if let id = eventField(event.payload, "labelId"), let name = labels?[id]?.name {
            return "removed label \(name)"
        }
        return "removed a label"
    case "pr_opened":
        if let n = eventField(event.payload, "prNumber") { return "opened PR #\(n)" }
        return "opened a pull request"
    case "pr_merged":
        if let n = eventField(event.payload, "prNumber") { return "merged PR #\(n)" }
        return "merged the pull request"
    default:
        return agentEventVerb(event.type)
    }
}
